File: services/k8s-service-go/internal/handler/security.go
package handler

import (
	"net/http"

	"github.com/go-chi/chi/v5"
	"github.com/junginho0901/kubeast/services/pkg/response"
)

// --- ServiceAccounts ---

// GetAllServiceAccounts handles GET /api/v1/serviceaccounts/all.
func (h *Handler) GetAllServiceAccounts(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	data, err := h.svc.GetAllServiceAccounts(ctx)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, data)
}

// GetServiceAccounts handles GET /api/v1/namespaces/{namespace}/serviceaccounts.
func (h *Handler) GetServiceAccounts(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	data, err := h.svc.GetServiceAccounts(ctx, namespace)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, data)
}

// DescribeServiceAccount handles GET /api/v1/namespaces/{namespace}/serviceaccounts/{name}/describe.
func (h *Handler) DescribeServiceAccount(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	name := chi.URLParam(r, "name")
	data, err := h.svc.DescribeServiceAccount(ctx, namespace, name)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, data)
}

// GetServiceAccountYAML handles GET /api/v1/namespaces/{namespace}/serviceaccounts/{name}/yaml.
func (h *Handler) GetServiceAccountYAML(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	name := chi.URLParam(r, "name")
	force := queryParamBool(r, "force_refresh", false)
	data, err := h.svc.GetGenericResourceYAML(ctx, "serviceaccounts", namespace, name, force)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, map[string]interface{}{"yaml": data})
}

// DeleteServiceAccount handles DELETE /api/v1/namespaces/{namespace}/serviceaccounts/{name}.
func (h *Handler) DeleteServiceAccount(w http.ResponseWriter, r *http.Request) {
	if err := h.requirePermission(r, "resource.serviceaccount.delete"); err != nil {
		h.handleError(w, err)
		return
	}
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	name := chi.URLParam(r, "name")
	err := h.svc.DeleteServiceAccount(ctx, namespace, name)
	h.recordAudit(r, "k8s.serviceaccount.delete", "serviceaccount", name, namespace, err)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, map[string]interface{}{"deleted": true})
}

// --- Roles ---

// GetAllRoles handles GET /api/v1/roles/all.
func (h *Handler) GetAllRoles(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	data, err := h.svc.GetAllRoles(ctx)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, data)
}

// GetRoles handles GET /api/v1/namespaces/{namespace}/roles.
func (h *Handler) GetRoles(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	data, err := h.svc.GetRoles(ctx, namespace)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, data)
}

// DescribeRole handles GET /api/v1/namespaces/{namespace}/roles/{name}/describe.
func (h *Handler) DescribeRole(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	name := chi.URLParam(r, "name")
	data, err := h.svc.DescribeRole(ctx, namespace, name)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, data)
}

// GetRoleYAML handles GET /api/v1/namespaces/{namespace}/roles/{name}/yaml.
func (h *Handler) GetRoleYAML(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	name := chi.URLParam(r, "name")
	force := queryParamBool(r, "force_refresh", false)
	data, err := h.svc.GetGenericResourceYAML(ctx, "roles", namespace, name, force)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, map[string]interface{}{"yaml": data})
}

// DeleteRole handles DELETE /api/v1/namespaces/{namespace}/roles/{name}.
func (h *Handler) DeleteRole(w http.ResponseWriter, r *http.Request) {
	if err := h.requirePermission(r, "resource.role.delete"); err != nil {
		h.handleError(w, err)
		return
	}
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	name := chi.URLParam(r, "name")
	err := h.svc.DeleteRole(ctx, namespace, name)
	h.recordAudit(r, "k8s.role.delete", "role", name, namespace, err)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, map[string]interface{}{"deleted": true})
}

// --- RoleBindings ---

// GetAllRoleBindings handles GET /api/v1/rolebindings/all.
func (h *Handler) GetAllRoleBindings(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	data, err := h.svc.GetAllRoleBindings(ctx)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, data)
}

// GetRoleBindings handles GET /api/v1/namespaces/{namespace}/rolebindings.
func (h *Handler) GetRoleBindings(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	data, err := h.svc.GetRoleBindings(ctx, namespace)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, data)
}

// DescribeRoleBinding handles GET /api/v1/namespaces/{namespace}/rolebindings/{name}/describe.
func (h *Handler) DescribeRoleBinding(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	name := chi.URLParam(r, "name")
	data, err := h.svc.DescribeRoleBinding(ctx, namespace, name)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, data)
}

// GetRoleBindingYAML handles GET /api/v1/namespaces/{namespace}/rolebindings/{name}/yaml.
func (h *Handler) GetRoleBindingYAML(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	name := chi.URLParam(r, "name")
	force := queryParamBool(r, "force_refresh", false)
	data, err := h.svc.GetGenericResourceYAML(ctx, "rolebindings", namespace, name, force)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, map[string]interface{}{"yaml": data})
}

// DeleteRoleBinding handles DELETE /api/v1/namespaces/{namespace}/rolebindings/{name}.
func (h *Handler) DeleteRoleBinding(w http.ResponseWriter, r *http.Request) {
	if err := h.requirePermission(r, "resource.rolebinding.delete"); err != nil {
		h.handleError(w, err)
		return
	}
	ctx := r.Context()
	namespace := chi.URLParam(r, "namespace")
	name := chi.URLParam(r, "name")
	err := h.svc.DeleteRoleBinding(ctx, namespace, name)
	h.recordAudit(r, "k8s.rolebinding.delete", "rolebinding", name, namespace, err)
	if err != nil {
		h.handleError(w, err)
		return
	}
	response.JSON(w, http.StatusOK, map[string]interface{}{"deleted": true})
}

File: frontend/src/services/api/security.ts
// Security API — RBAC objects: ServiceAccounts / Roles / RoleBindings
// / ClusterRoles / ClusterRoleBindings. PodSecurityPolicies are not
// served by the backend any more (deprecated upstream).

import { client } from './client'
import type {
  ClusterRoleBindingInfo,
  ClusterRoleInfo,
  RoleBindingInfo,
  RoleInfo,
  ServiceAccountInfo,
} from './types'

export const securityApi = {
  // ServiceAccounts
  getServiceAccounts: async (namespace: string, forceRefresh = false): Promise<ServiceAccountInfo[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/serviceaccounts`, {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  getAllServiceAccounts: async (forceRefresh = false): Promise<ServiceAccountInfo[]> => {
    const { data } = await client.get('/cluster/serviceaccounts/all', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeServiceAccount: async (namespace: string, name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/serviceaccounts/${name}/describe`)
    return data
  },

  deleteServiceAccount: async (namespace: string, name: string): Promise<void> => {
    await client.delete(`/cluster/namespaces/${namespace}/serviceaccounts/${name}`)
  },

  // Roles
  getRoles: async (namespace: string, forceRefresh = false): Promise<RoleInfo[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/roles`, {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  getAllRoles: async (forceRefresh = false): Promise<RoleInfo[]> => {
    const { data } = await client.get('/cluster/roles/all', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeRole: async (namespace: string, name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/roles/${name}/describe`)
    return data
  },

  deleteRole: async (namespace: string, name: string): Promise<void> => {
    await client.delete(`/cluster/namespaces/${namespace}/roles/${name}`)
  },

  // RoleBindings
  getRoleBindings: async (namespace: string, forceRefresh = false): Promise<RoleBindingInfo[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/rolebindings`, {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  getAllRoleBindings: async (forceRefresh = false): Promise<RoleBindingInfo[]> => {
    const { data } = await client.get('/cluster/rolebindings/all', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeRoleBinding: async (namespace: string, name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/rolebindings/${name}/describe`)
    return data
  },

  deleteRoleBinding: async (namespace: string, name: string): Promise<void> => {
    await client.delete(`/cluster/namespaces/${namespace}/rolebindings/${name}`)
  },

  // ClusterRoles
  getClusterRoles: async (forceRefresh = false): Promise<ClusterRoleInfo[]> => {
    const { data } = await client.get('/cluster/clusterroles', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeClusterRole: async (name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/clusterroles/${name}/describe`)
    return data
  },

  deleteClusterRole: async (name: string): Promise<void> => {
    await client.delete(`/cluster/clusterroles/${name}`)
  },

  // ClusterRoleBindings
  getClusterRoleBindings: async (forceRefresh = false): Promise<ClusterRoleBindingInfo[]> => {
    const { data } = await client.get('/cluster/clusterrolebindings', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeClusterRoleBinding: async (name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/clusterrolebindings/${name}/describe`)
    return data
  },

  deleteClusterRoleBinding: async (name: string): Promise<void> => {
    await client.delete(`/cluster/clusterrolebindings/${name}`)
  },
}
